File: Csharp/SolidCourse/LSP/Before.cs
// Base class
public class Bird
{
    public virtual void Walk()
    {
        Console.WriteLine("I can walk!");
    }
    public virtual void Fly()
    {
        Console.WriteLine("I can fly!");
    }
}

// Derived class
public class Eagle : Bird
{
    public override void Walk()
    {
        Console.WriteLine("Eagle can walk!");
    }
    public override void Fly()
    {
        Console.WriteLine("Sparrow flying high!");
    }
}

// Another derived class
public class Penguin : Bird
{
    public override void Walk()
    {
        Console.WriteLine("Penguin can walk!");
    }
    public override void Fly()
    {
        throw new NotSupportedException("Penguins cannot fly.");
    }
}

// Client Code
public class BirdManager
{
    public void MakeBirdFly(Bird bird)
    {
        bird.Fly(); // This works for Eagle but not for Penguin
    }
}

File: CPlusPlus/LSP/Before.cpp
#include <iostream>
#include <stdexcept>
#include <memory>

class Bird {
public:
    virtual void Walk() {
        std::cout << "I can walk!" << std::endl;
    }

    virtual void Fly() {
        std::cout << "I can fly!" << std::endl;
    }

    virtual ~Bird() = default;
};

class Eagle : public Bird {
public:
    void Walk() override {
        std::cout << "Eagle can walk!" << std::endl;
    }

    void Fly() override {
        std::cout << "Sparrow flying high!" << std::endl;
    }
};

class Penguin : public Bird {
public:
    void Walk() override {
        std::cout << "Penguin can walk!" << std::endl;
    }

    void Fly() override {
        throw std::logic_error("Penguins cannot fly.");
    }
};

class BirdManager {
public:
    void MakeBirdFly(Bird& bird) {
        bird.Fly(); // Throws for Penguin
    }
};
